Pinta.Core/Effects/ColorGradient.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Pinta.Core;
+
+public sealed class ColorGradient : ColorMapping
+{
+	public ColorBgra StartColor { get; }
+	public ColorBgra EndColor { get; }
+	public double MinimumPosition { get; }
+	public double MaximumPosition { get; }
+
+	private readonly IReadOnlyList<GradientStop> sorted_stops; // TODO: Before making it public think of nice structures
+
+	internal ColorGradient (
+		ColorBgra startColor,
+		ColorBgra endColor,
+		double minPosition,
+		double maxPosition,
+		IEnumerable<KeyValuePair<double, ColorBgra>> gradientStops)
+	{
+		if (minPosition >= maxPosition) throw new ArgumentException ($"{nameof (minPosition)} has to be lower than {nameof (maxPosition)}");
+		var sortedStops = gradientStops.Select (kvp => new GradientStop (kvp.Key, kvp.Value)).OrderBy (stop => stop.Position).ToArray ();
+		if (sortedStops.Length > 0 && sortedStops[0].Position <= minPosition) throw new ArgumentException ($"Lowest key in {nameof (gradientStops)} has to be greater than {nameof (minPosition)}");
+		if (sortedStops.Length > 0 && sortedStops[^1].Position >= maxPosition) throw new ArgumentException ($"Greatest key in {nameof (gradientStops)} has to be lower than {nameof (maxPosition)}");
+		var distinctPositions = sortedStops.GroupBy (s => s.Position).Count ();
+		if (distinctPositions != sortedStops.Length) throw new ArgumentException ("Cannot have more than one stop in the same position");
+
+		StartColor = startColor;
+		EndColor = endColor;
+		MinimumPosition = minPosition;
+		MaximumPosition = maxPosition;
+		sorted_stops = sortedStops;
+	}
+
+	public sealed override ColorBgra GetColor (double position)
+	{
+		if (position == MinimumPosition) return StartColor;
+		if (position == MaximumPosition) return EndColor;
+		if (!IsMapped (position)) throw new ArgumentOutOfRangeException (nameof (position));
```

Comment:
If the position is out of range, perhaps this should just clamp to the start / end color? Otherwise this requires the caller to always carefully clamp which might be unexpected

Reply:
Do you think clamping would be better, or adjusting the start/end color to be an intermediate between the out-of-range color and the next stop?

Reply:
I think just clamping makes sense, since there's nothing well-defined before the start color to continue blending into.
One other thing I just noticed too is that the effects currently assume that the gradient has a 0-1024 range, which might not actually be the case for whatever mapping the effect is provided. So that's also something to think about with the API, e.g. maybe it's easiest for the mapping to work in a 0-1 range and the effect remaps its 0-1024 value to that